alchemist-full/build.gradle.kts

```diff
@@ -30,6 +37,99 @@ application {
+    input = rootProject.layout.buildDirectory.map { it.dir("package-input") }.get().asFile.path
+    destination = rootProject.layout.buildDirectory.map { it.dir("package") }.get().asFile.path
+    mainJar = "alchemist-full-${rootProject.version}-all.jar"
+    mainClass = "it.unibo.alchemist.Alchemist"
+
+    linux {
+        icon = "${rootProject.projectDir}/package-settings/logo.png"
+        type = ImageType.RPM
+    }
+    windows {
+        icon = "${rootProject.projectDir}/package-settings/logo.ico"
+        type = ImageType.MSI
+        winDirChooser = true
+        winShortcutPrompt = true
+        winPerUserInstall = isInCI
+    }
+    mac {
+        icon = "${rootProject.projectDir}/package-settings/logo.png"
+        type = ImageType.PKG
+    }
+
+    dependsOn("copyForPackaging")
+}
+
+tasks.register<Copy>("copyForPackaging") {
+    val jarFile = "alchemist-full-${rootProject.version}-all.jar"
+    from("${rootProject.projectDir}/build/shadow/$jarFile")
+    into("${rootProject.projectDir}/build/package-input")
+    dependsOn("shadowJar")
+}
+
+tasks.register<Exec>("testJpackageOutput") {
+    group = "Verification"
+    description = "Verifies the jpackage output correctness for the OS running the script"
+    val interceptOutput = ByteArrayOutputStream()
+    val interceptError = ByteArrayOutputStream()
+    val version = rootProject.version.toString().substringBefore('-')
+    val isLinux = isUnix && !isMac
+    standardOutput = interceptOutput
+    errorOutput = interceptError
+    isIgnoreExitValue = true
+    workingDir = rootProject.file("build/package/")
+    doFirst {
+        // Extract the packet
+        if (isWindows) {
+            commandLine("msiexec", "-i", "${rootProject.name}-$version.msi", "-quiet", "INSTALLDIR=${workingDir.path}\\install")
+        } else if (isMac) {
+            commandLine("sudo", "installer", "-pkg", "${rootProject.name}-$version.pkg", "-target", "/")
+        } else if (isLinux) {
+            workingDir.resolve("install").mkdirs()
+            commandLine("bsdtar", "-xf", "${rootProject.name}-$version-1.x86_64.rpm", "-C", "install")
+        }
+    }
+    doLast {
+        // Check if package contains every file needed
+        var execFiles = listOf("")
+        var appFiles = listOf("")
+        if (isWindows) {
+            execFiles = workingDir.resolve("install").listFiles().map { it.name }
+            appFiles = workingDir.resolve("install/app").listFiles().map { it.name }
+        } else if (isMac) {
+            val root = File("/Applications/${rootProject.name}.app")
+            execFiles = root.resolve("Contents/MacOS").listFiles().map { it.name }
+            appFiles = root.resolve("Contents/app").listFiles().map { it.name }
+        } else if (isLinux) {
+            execFiles = workingDir.resolve("install/opt/alchemist/bin").listFiles().map { it.name }
+            appFiles = workingDir.resolve("install/opt/alchemist/lib/app").listFiles().map { it.name }
+        }
```

Comment:
this looks very clumsy and copy/pasted. Why does MacOS require a position in the file system root, while the others don't?

Reply:
In Windows and Linux using "msiexec" and "bsdtar" we install/extract in a specific directory ("build/package/install") while in MacOS I did not find a way to tell the "installer" where to install the package. That is why the package is automatically installed in the Applications/ directory of the vm.

Reply:
We could set the installation directory from the jpackage configuration: 
 - If you run the testJpackageOutput task, the jpackage task will run with an arbitrary installation directory
 - If you run the jpackage task directly it will use the standard path

Reply:
This is unclear to me. The task should produce an installable package. The installable package should be capable to install without Gradle being involved on arbitrary directories, or, in case of Linux, have pre-defined absolute paths.
Here, I see two relative and two absolute paths.

It's not something related to build, I need to understand *exactly* what these parameters are for.

Reply:
> The installation directory is platform-specific:
>  - On Linux, the default is /opt/application-name
>  - On macOS, the default is /Applications/application-name
>  - On Windows, the default is c:\Program Files\application-name; if the --win-per-user-install option is used, the default is C:\Users\user-name\AppData\Local\application-name
[source](https://docs.oracle.com/en/java/javase/21/jpackage/basic-packaging.html#GUID-EB1303BE-292F-4FEE-B91E-3C9CC367324B)


In the current configuration, every installable package is set to the default path. Windows has the property `winDirChooser = true`. Consequently, when you double-click the installer, it opens the classic installation GUI that allows you to modify the path where the application will be installed.

In the test task, I intentionally enforced the installation into a temporary "install" directory in the following ways:
- On Windows, by setting the variable INSTALLDIR.
- On Linux, by using the -C option in `bsdtar`.
- On MacOS, I could not find a way to force the installation into a custom directory. Therefore, the `installer` command automatically installs the package in `/Applications/alchemist`. 

I opted for using a temporary directory solely for convenience.
I hope this clarification makes everything clearer.